File: Algorithms/1022.Sum-of-Root-To-Leaf-Binary-Numbers/sum_of_root_to_leaf_binary_numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sumRootToLeaf(self, root: TreeNode) -> int:
        if not root:
            return 0
        self.res = 0
        self.dfs(root, root.val)
        return self.res
    
    
    def dfs(self, root, curSum):
        if not root.left and not root.right:
            self.res = self.res + curSum
            return
        if root.left:
            self.dfs(root.left, curSum * 2 + root.left.val)
        if root.right:
            self.dfs(root.right, curSum * 2 + root.right.val)

File: Algorithms/1022.Sum-of-Root-To-Leaf-Binary-Numbers/sum_of_root_to_leaf_binary_numbers.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumRootToLeaf(root *TreeNode) int {
    return dfs(root, 0)
}

func dfs(root *TreeNode, cur int) int {
    if root == nil {
        return 0
    }
    
    cur = cur * 2 + root.Val
    if root.Left == nil && root.Right == nil {
        return cur
    }
    return dfs(root.Left, cur) + dfs(root.Right, cur)
}